src/metax_api/services/file_service.py

```diff
@@ -812,44 +811,45 @@ def _get_directory_contents_for_catalog_record(directory_id, cr_id, not_cr_id, d
         # dirs which otherwise contained files, but not any files that were selected for
         # the cr, are not returned.
 
+        directory_fields_string = ', '.join([field.replace('parent_directory__', 'parent_d.')
+            if 'parent_directory__' in field else 'd.' + field for field in directory_fields])
+
         sql_select_dirs_for_cr = """
-            select d.id
-            from metax_api_directory d
-            where d.parent_directory_id = {}
-            and exists(
-                select 1
-                from metax_api_file f
-                inner join metax_api_catalogrecord_files cr_f on cr_f.file_id = f.id
-                where f.file_path like (d.directory_path || '/%%')
-                and cr_f.catalogrecord_id {} {}
-                and f.removed = false
-                and f.active = true
+            SELECT {}
+            FROM metax_api_directory d
+            JOIN metax_api_directory parent_d
+                ON d.parent_directory_id = parent_d.id
+            WHERE d.parent_directory_id = {}
+            AND EXISTS(
+                SELECT 1
+                FROM metax_api_file f
+                INNER JOIN metax_api_catalogrecord_files cr_f ON cr_f.file_id = f.id
+                WHERE f.file_path LIKE (d.directory_path || '/%%')
+                AND cr_f.catalogrecord_id {} {}
+                AND f.removed = false
+                AND f.active = true
             )
             """
 
         with connection.cursor() as cr:
-            if dirs_only:
-                files = []
-            else:
-                if cr_id:
-                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '=', cr_id))
-                    files = None if dirs_only else File.objects \
-                        .filter(record__pk=cr_id, parent_directory=directory_id, removed=False, active=True) \
-                        .values(*file_fields)
-                if not_cr_id:
-                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '!=', not_cr_id))
-                    files = None if dirs_only else File.objects.exclude(record__pk=not_cr_id) \
-                        .filter(parent_directory=directory_id, removed=False, active=True) \
-                        .values(*file_fields)
-            directory_ids = [ row[0] for row in cr.fetchall() ]
-
-        if not directory_ids and not files:
+            if cr_id:
+                cr.execute(sql_select_dirs_for_cr.format(directory_fields_string, directory_id, '=', cr_id))
```

Comment:
Using string ``format`` here, with line 814 where directory fields are formed by concatenating string that are received from a query parameter, is again somewhat hazardous.

I think it would be worthy of leaving a comment in the code that using this string ``format`` here is safe only by the grace of ``_get_requested_file_browsing_fields()`` which is called somewhere earlier during this HTTP request, and which checks that the given fields are withing the list of allowed directory fields as defined in the DirectorySerializer. But even then, I would research if forming raw SQL-queries dynamically can benefit from passing the field names as parameters to ``cr.execute()`` instead of using ``format()``.

Further reading:
- https://docs.djangoproject.com/en/2.2/topics/security/#sql-injection-protection
- https://docs.djangoproject.com/en/2.2/topics/db/sql/#passing-parameters-into-raw
- https://docs.djangoproject.com/en/2.2/topics/db/sql/#executing-custom-sql-directly

``cr.execute()`` has features to prevent SQL-injections, so those should be used whenever possible (or if not used, leave a comment in the code why not).

Reply:
Noticed that parameter passing was used in a following commit, which is good. Still worth looking into if the directory field names can be passed in the same manner.

Reply:
Tried to make requests to directory api with different inputs for directory_fields and file_fields and it is pretty well  secured with allowed_fields check. Is it really possible to overcome this validation in LightDirectorySerializer ?

Reply:
Yes, the ``allowed_fields`` check done previously in the code makes it safe. I am just being extra cautious for future changes elsewhere that may not take into account that the safety of this ``cr.execute()`` depends on an earlier step that may not be so obvious, since this is a public API endpoint and an oversight here, or in the code where the allowed_fields check is done, can open a way for injections.

Even better than leaving an additional comment in the code, maybe it would be a good idea to do an additional check that all the fields belong to ``serializer.allowed_fields`` (maybe around line 814), so that the safety of this cr.execute() does not depend on a much earlier piece of code, which may change at some point in time without taking into account that it is actually the only step that makes this part safe to use?

This may sound overly cautious, but I think this is one of the only places (if not the only place?) where sql-query parameters are sanitized elsewhere than where the query is executed, so an unrelated change elsewhere (or a bug which for whatever reason skips the earlier field checking) can affect the safety of this query.